library/core/src/fmt/num.rs

```diff
@@ -199,44 +200,58 @@ static DEC_DIGITS_LUT: &[u8; 200] = b"\
       6061626364656667686970717273747576777879\
       8081828384858687888990919293949596979899";
 
-macro_rules! impl_Display {
-    ($($signed:ident, $unsigned:ident,)* ; as $u:ident via $conv_fn:ident named $gen_name:ident) => {
+static NEGATIVE_SIGN: &[u8; 1] = b"-";
+
+// SAFETY: safety is ensured by the caller about:
+// 1. The contents of `buf` containing only ASCII characters.
+// 2. `offset` being bound checked.
+// 3. `buffer.contents` being initialized from `offset` onwards till the end
+unsafe fn _extract_str_from_buf(buf: &NumBuffer, offset: usize) -> &str {
+    // SAFETY: safety is ensured by the caller about:
+    // 1. `offset` being bound checked
+    // 2. `buffer.contents` being initialized from `offset` onwards
+    let written = unsafe { buf.extract(offset..) };
+
+    // SAFETY: safety is ensured by the caller about:
+    // 1. The contents of `buf` containing only ASCII characters
+    let as_str = unsafe {
+        str::from_utf8_unchecked(slice::from_raw_parts(
+            MaybeUninit::slice_as_ptr(written),
+            written.len(),
+        ))
+    };
 
-        $(
-        #[stable(feature = "rust1", since = "1.0.0")]
-        impl fmt::Display for $unsigned {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                #[cfg(not(feature = "optimize_for_size"))]
-                {
-                    self._fmt(true, f)
-                }
-                #[cfg(feature = "optimize_for_size")]
-                {
-                    $gen_name(self.$conv_fn(), true, f)
-                }
-            }
-        }
+    as_str
+}
 
-        #[stable(feature = "rust1", since = "1.0.0")]
-        impl fmt::Display for $signed {
-            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-                #[cfg(not(feature = "optimize_for_size"))]
-                {
-                    return self.unsigned_abs()._fmt(*self >= 0, f);
-                }
-                #[cfg(feature = "optimize_for_size")]
-                {
-                    return $gen_name(self.unsigned_abs().$conv_fn(), *self >= 0, f);
-                }
-            }
-        }
+// SAFETY: safety is ensured by the caller about:
+// 1. `start_offset` being bound checked
+unsafe fn _add_negative_sign(
```

Comment:
```suggestion
unsafe fn add_negative_sign(
```
Free functions don't need the underscore prefix, `_fmt` only has it because it's an inherent method and would clash with the `fmt` methods of the formatting traits (Display, Debug, etc.).

Reply:
Got it